dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java

```diff
@@ -474,27 +479,131 @@ public JdbcEventStore( StatementBuilder statementBuilder, @Qualifier( "readOnlyJ
         return eventRows;
     }
 
-    private Map<String, String> getDataElementsUidToCode( Set<EventDataValue> eventDataValues, IdScheme idScheme )
+    private Map<String, String> getDataElementsUidToIdentifier( Set<EventDataValue> eventDataValues, IdScheme idScheme )
+    {
+        if ( idScheme.isAttribute() )
+        {
+            return getDataElementsUidToAttribute( eventDataValues );
+        }
+        else
+        {
+            return getDataElementsUidToCode( eventDataValues );
+        }
+    }
+
+    private Map<String, String> getDataElementsUidToCode( Set<EventDataValue> eventDataValues )
     {
         //Get mapping only when needed
-        if ( idScheme != IdScheme.ID && idScheme != IdScheme.UID )
+        Map<String, String> dataElementsUidToCode = new HashMap<>();
+
+        List<String> dataElementUids = eventDataValues.stream()
+            .map( EventDataValue::getDataElement )
+            .collect( Collectors.toList() );
+
```

Comment:
Instead of fetching this in Java (slow!) could we instead fetch this in SQL, e.g. return the code of the data elements in the SQL instead of UID?

Reply:
Again, as discussed in Slack. I am not sure there is an easy way how to do that in SQL as EventDataValues are kept in JSONB. But can definitely investigate it.

Reply:
My investigation is that it won't be trivial. And in the end, I am not sure we will improve the performance. We will make the code more complex (especially the SQL query) and we will also have to return more rows (something similar to what the solution was in 2.31 and below) and process them in memory. So we will improve performance in one place but make it worse another place.

Reply:
What about something like
`select uid,(array_to_json(ARRAY(SELECT jsonb_object_keys(eventdatavalues)))) as dataElements from programstageinstance;`

Might be possible to remove the ARRAY part also, but wouldn't it be quicker than doing it in java?

Reply:
I believe the issue that bothers Lars is not how we currently fetch UIDs, but how we currently fetch Codes (this line below: DataElement de = dataElementToCodeCache.get( uid, dataElementCallable.setId( uid ) );) and whether we could fetch them directly in SQL query.

The problem is that in `eventdatavalues` we use UID as a key and we do not keep track of Code. Therefore, obtaining Code of the DataElement would be quite complex. We would have to "flatmap" keys (UIDs) from `eventdatavalues` and join each one with `dataelement` table on UID in order to obtain Code. That would create many new joins and rows (based on number of data values in `eventdatavalues`). And then we would have to process these many more rows in the code (something similar how we were processing rows in 2.31, when we still used `trackedentitydatavalue` table - https://github.com/dhis2/dhis2-core/blob/5619eeaddcc24445265f4d017d44e12c377924bd/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java#L159

Reply:
Wouldn't that at least be a perfect candidate for a `SELECT IN(...)` for ? to get all data elements